File: csharp/src/Fory/PrimitiveSerializers.cs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

namespace Apache.Fory;

internal enum ForyStringEncoding : ulong
{
    Latin1 = 0,
    Utf16 = 1,
    Utf8 = 2,
}

public sealed class BoolSerializer : Serializer<bool>
{

    public override bool DefaultValue => false;

    public override void WriteData(WriteContext context, in bool value, bool hasGenerics)
    {
        _ = hasGenerics;
        context.Writer.WriteUInt8(value ? (byte)1 : (byte)0);
    }

    public override bool ReadData(ReadContext context)
    {
        return context.Reader.ReadUInt8() != 0;
    }
}

public sealed class Int8Serializer : Serializer<sbyte>
{

    public override sbyte DefaultValue => 0;

    public override void WriteData(WriteContext context, in sbyte value, bool hasGenerics)
    {
        _ = hasGenerics;
        context.Writer.WriteInt8(value);
    }

    public override sbyte ReadData(ReadContext context)
    {
        return context.Reader.ReadInt8();
    }
}

public sealed class Int16Serializer : Serializer<short>
{

    public override short DefaultValue => 0;

    public override void WriteData(WriteContext context, in short value, bool hasGenerics)
    {
        _ = hasGenerics;
        context.Writer.WriteInt16(value);
    }

    public override short ReadData(ReadContext context)
    {
        return context.Reader.ReadInt16();
    }
}

public sealed class Int32Serializer : Serializer<int>
{

    public override int DefaultValue => 0;

    public override void WriteData(WriteContext context, in int value, bool hasGenerics)
    {
        _ = hasGenerics;
        context.Writer.WriteVarInt32(value);
    }

    public override int ReadData(ReadContext context)
    {
        return context.Reader.ReadVarInt32();
    }
}

public sealed class Int64Serializer : Serializer<long>
{

    public override long DefaultValue => 0;

    public override void WriteData(WriteContext context, in long value, bool hasGenerics)
    {
        _ = hasGenerics;
        context.Writer.WriteVarInt64(value);
    }

    public override long ReadData(ReadContext context)
    {
        return context.Reader.ReadVarInt64();
    }
}

public sealed class UInt8Serializer : Serializer<byte>
{

    public override byte DefaultValue => 0;

    public override void WriteData(WriteContext context, in byte value, bool hasGenerics)
    {
        _ = hasGenerics;
        context.Writer.WriteUInt8(value);
    }

    public override byte ReadData(ReadContext context)
    {
        return context.Reader.ReadUInt8();
    }
}

public sealed class UInt16Serializer : Serializer<ushort>
{

    public override ushort DefaultValue => 0;

    public override void WriteData(WriteContext context, in ushort value, bool hasGenerics)
    {
        _ = hasGenerics;
        context.Writer.WriteUInt16(value);
    }

    public override ushort ReadData(ReadContext context)
    {
        return context.Reader.ReadUInt16();
    }
}

public sealed class UInt32Serializer : Serializer<uint>
{

    public override uint DefaultValue => 0;

    public override void WriteData(WriteContext context, in uint value, bool hasGenerics)
    {
        _ = hasGenerics;
        context.Writer.WriteVarUInt32(value);
    }

    public override uint ReadData(ReadContext context)
    {
        return context.Reader.ReadVarUInt32();
    }
}

public sealed class UInt64Serializer : Serializer<ulong>
{

    public override ulong DefaultValue => 0;

    public override void WriteData(WriteContext context, in ulong value, bool hasGenerics)
    {
        _ = hasGenerics;
        context.Writer.WriteVarUInt64(value);
    }

    public override ulong ReadData(ReadContext context)
    {
        return context.Reader.ReadVarUInt64();
    }
}

public sealed class Float16Serializer : Serializer<Half>
{

    public override Half DefaultValue => default;

    public override void WriteData(WriteContext context, in Half value, bool hasGenerics)
    {
        _ = hasGenerics;
        context.Writer.WriteUInt16(BitConverter.HalfToUInt16Bits(value));
    }

    public override Half ReadData(ReadContext context)
    {
        return BitConverter.UInt16BitsToHalf(context.Reader.ReadUInt16());
    }
}

public sealed class BFloat16Serializer : Serializer<BFloat16>
{

    public override BFloat16 DefaultValue => default;

    public override void WriteData(WriteContext context, in BFloat16 value, bool hasGenerics)
    {
        _ = hasGenerics;
        context.Writer.WriteUInt16(value.ToBits());
    }

    public override BFloat16 ReadData(ReadContext context)
    {
        return BFloat16.FromBits(context.Reader.ReadUInt16());
    }
}

public sealed class Float32Serializer : Serializer<float>
{

    public override float DefaultValue => 0;

    public override void WriteData(WriteContext context, in float value, bool hasGenerics)
    {
        _ = hasGenerics;
        context.Writer.WriteFloat32(value);
    }

    public override float ReadData(ReadContext context)
    {
        return context.Reader.ReadFloat32();
    }
}

public sealed class Float64Serializer : Serializer<double>
{

    public override double DefaultValue => 0;

    public override void WriteData(WriteContext context, in double value, bool hasGenerics)
    {
        _ = hasGenerics;
        context.Writer.WriteFloat64(value);
    }

    public override double ReadData(ReadContext context)
    {
        return context.Reader.ReadFloat64();
    }
}

public sealed class BinarySerializer : Serializer<byte[]>
{


    public override byte[] DefaultValue => null!;

    public override void WriteData(WriteContext context, in byte[] value, bool hasGenerics)
    {
        _ = hasGenerics;
        byte[] safe = value ?? [];
        context.Writer.WriteVarUInt32((uint)safe.Length);
        context.Writer.WriteBytes(safe);
    }

    public override byte[] ReadData(ReadContext context)
    {
        uint length = context.Reader.ReadVarUInt32();
        return context.Reader.ReadBytes(checked((int)length));
    }
}

File: dart/packages/fory/lib/src/serializer/primitive_serializers.dart
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import 'package:fory/src/context/read_context.dart';
import 'package:fory/src/context/write_context.dart';
import 'package:fory/src/meta/type_ids.dart';
import 'package:fory/src/serializer/serializer.dart';
import 'package:fory/src/types/bfloat16.dart';
import 'package:fory/src/types/float16.dart';
import 'package:fory/src/types/float32.dart';
import 'package:fory/src/types/int16.dart';
import 'package:fory/src/types/int32.dart';
import 'package:fory/src/types/int64.dart';
import 'package:fory/src/types/int8.dart';
import 'package:fory/src/types/uint16.dart';
import 'package:fory/src/types/uint32.dart';
import 'package:fory/src/types/uint64.dart';
import 'package:fory/src/types/uint8.dart';

const int _jsSafeUint64IntMax = 9007199254740991;
const bool _isWeb = bool.fromEnvironment('dart.library.js_interop');

final class PrimitiveSerializer<T> extends Serializer<T> {
  final int typeId;
  final bool _supportsRef;

  const PrimitiveSerializer(
    this.typeId, {
    required bool supportsRef,
  }) : _supportsRef = supportsRef;

  @override
  bool get supportsRef => _supportsRef;

  @override
  void write(WriteContext context, T value) {
    writePayload(context, typeId, value as Object);
  }

  @override
  T read(ReadContext context) {
    return readPayload(context, typeId) as T;
  }

  static void writePayload(
    WriteContext context,
    int typeId,
    Object value,
  ) {
    final buffer = context.buffer;
    switch (typeId) {
      case TypeIds.boolType:
        buffer.writeBool(value as bool);
        return;
      case TypeIds.int8:
        buffer.writeByte((value as Int8).value);
        return;
      case TypeIds.int16:
        buffer.writeInt16((value as Int16).value);
        return;
      case TypeIds.int32:
        buffer.writeInt32((value as Int32).value);
        return;
      case TypeIds.varInt32:
        buffer.writeVarInt32(value is Int32 ? value.value : value as int);
        return;
      case TypeIds.int64:
        if (value is Int64) {
          buffer.writeInt64(value);
        } else {
          buffer.writeInt64FromInt(value as int);
        }
        return;
      case TypeIds.varInt64:
        if (value is Int64) {
          buffer.writeVarInt64(value);
        } else {
          buffer.writeVarInt64FromInt(value as int);
        }
        return;
      case TypeIds.taggedInt64:
        if (value is Int64) {
          buffer.writeTaggedInt64(value);
        } else {
          buffer.writeTaggedInt64FromInt(value as int);
        }
        return;
      case TypeIds.uint8:
        buffer.writeUint8(value is Uint8 ? value.value : value as int);
        return;
      case TypeIds.uint16:
        buffer.writeUint16(value is Uint16 ? value.value : value as int);
        return;
      case TypeIds.uint32:
        buffer.writeUint32(value is Uint32 ? value.value : value as int);
        return;
      case TypeIds.varUint32:
        buffer.writeVarUint32(value is Uint32 ? value.value : value as int);
        return;
      case TypeIds.uint64:
        buffer.writeUint64(_uint64Value(value));
        return;
      case TypeIds.varUint64:
        buffer.writeVarUint64(_uint64Value(value));
        return;
      case TypeIds.taggedUint64:
        buffer.writeTaggedUint64(_uint64Value(value));
        return;
      case TypeIds.float16:
        buffer.writeFloat16(value as Float16);
        return;
      case TypeIds.bfloat16:
        buffer.writeBfloat16(value as Bfloat16);
        return;
      case TypeIds.float32:
        buffer.writeFloat32((value as Float32).value);
        return;
      case TypeIds.float64:
        buffer.writeFloat64(value as double);
        return;
      default:
        throw StateError('Unsupported primitive type id $typeId.');
    }
  }

  static Object readPayload(
    ReadContext context,
    int typeId,
  ) {
    final buffer = context.buffer;
    switch (typeId) {
      case TypeIds.boolType:
        return buffer.readBool();
      case TypeIds.int8:
        return Int8(buffer.readByte());
      case TypeIds.int16:
        return Int16(buffer.readInt16());
      case TypeIds.int32:
        return Int32(buffer.readInt32());
      case TypeIds.varInt32:
        return Int32(buffer.readVarInt32());
      case TypeIds.int64:
        return buffer.readInt64();
      case TypeIds.varInt64:
        return buffer.readVarInt64();
      case TypeIds.taggedInt64:
        return buffer.readTaggedInt64();
      case TypeIds.uint8:
        return Uint8(buffer.readUint8());
      case TypeIds.uint16:
        return Uint16(buffer.readUint16());
      case TypeIds.uint32:
        return Uint32(buffer.readUint32());
      case TypeIds.varUint32:
        return Uint32(buffer.readVarUint32());
      case TypeIds.uint64:
        return buffer.readUint64();
      case TypeIds.varUint64:
        return buffer.readVarUint64();
      case TypeIds.taggedUint64:
        return buffer.readTaggedUint64();
      case TypeIds.float16:
        return buffer.readFloat16();
      case TypeIds.bfloat16:
        return buffer.readBfloat16();
      case TypeIds.float32:
        return Float32(buffer.readFloat32());
      case TypeIds.float64:
        return buffer.readFloat64();
      default:
        throw StateError('Unsupported primitive type id $typeId.');
    }
  }
}

Uint64 _uint64Value(Object value) {
  if (value is Uint64) {
    return value;
  }
  final intValue = value as int;
  if (_isWeb && (intValue < 0 || intValue > _jsSafeUint64IntMax)) {
    throw StateError(
      'Dart int value $intValue is outside the JS-safe unsigned uint64 '
      'int field range [0, $_jsSafeUint64IntMax]. Use Uint64 for full '
      'unsigned 64-bit values on web.',
    );
  }
  return Uint64(intValue);
}

const PrimitiveSerializer<bool> boolSerializer = PrimitiveSerializer<bool>(
  TypeIds.boolType,
  supportsRef: false,
);
const PrimitiveSerializer<Int8> int8Serializer = PrimitiveSerializer<Int8>(
  TypeIds.int8,
  supportsRef: false,
);
const PrimitiveSerializer<Int16> int16Serializer = PrimitiveSerializer<Int16>(
  TypeIds.int16,
  supportsRef: false,
);
const PrimitiveSerializer<Int32> int32Serializer = PrimitiveSerializer<Int32>(
  TypeIds.int32,
  supportsRef: false,
);
const PrimitiveSerializer<Int32> varInt32Serializer =
    PrimitiveSerializer<Int32>(
  TypeIds.varInt32,
  supportsRef: false,
);
const PrimitiveSerializer<Int64> int64Serializer = PrimitiveSerializer<Int64>(
  TypeIds.int64,
  supportsRef: false,
);
const PrimitiveSerializer<Int64> varInt64Serializer =
    PrimitiveSerializer<Int64>(
  TypeIds.varInt64,
  supportsRef: false,
);
const PrimitiveSerializer<Int64> taggedInt64Serializer =
    PrimitiveSerializer<Int64>(
  TypeIds.taggedInt64,
  supportsRef: false,
);
const PrimitiveSerializer<Uint8> uint8Serializer = PrimitiveSerializer<Uint8>(
  TypeIds.uint8,
  supportsRef: false,
);
const PrimitiveSerializer<Uint16> uint16Serializer =
    PrimitiveSerializer<Uint16>(
  TypeIds.uint16,
  supportsRef: false,
);
const PrimitiveSerializer<Uint32> uint32Serializer =
    PrimitiveSerializer<Uint32>(
  TypeIds.uint32,
  supportsRef: false,
);
const PrimitiveSerializer<Uint32> varUint32Serializer =
    PrimitiveSerializer<Uint32>(
  TypeIds.varUint32,
  supportsRef: false,
);
const PrimitiveSerializer<Uint64> uint64Serializer =
    PrimitiveSerializer<Uint64>(
  TypeIds.uint64,
  supportsRef: false,
);
const PrimitiveSerializer<Uint64> varUint64Serializer =
    PrimitiveSerializer<Uint64>(
  TypeIds.varUint64,
  supportsRef: false,
);
const PrimitiveSerializer<Uint64> taggedUint64Serializer =
    PrimitiveSerializer<Uint64>(
  TypeIds.taggedUint64,
  supportsRef: false,
);
const PrimitiveSerializer<Float16> float16Serializer =
    PrimitiveSerializer<Float16>(
  TypeIds.float16,
  supportsRef: false,
);
const PrimitiveSerializer<Bfloat16> bfloat16Serializer =
    PrimitiveSerializer<Bfloat16>(
  TypeIds.bfloat16,
  supportsRef: false,
);
const PrimitiveSerializer<Float32> float32Serializer =
    PrimitiveSerializer<Float32>(
  TypeIds.float32,
  supportsRef: false,
);
const PrimitiveSerializer<double> float64Serializer =
    PrimitiveSerializer<double>(
  TypeIds.float64,
  supportsRef: false,
);
